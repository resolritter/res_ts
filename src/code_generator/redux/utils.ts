import { snakeCase } from "lodash"

export function toReduxActionName(identifier: string): string {
  return snakeCase(identifier).toUpperCase()
}

