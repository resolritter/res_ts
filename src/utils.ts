import * as fs from "fs"
import * as path from "path"
import * as TJS from "typescript-json-schema"
import { JSONSchema7 } from "json-schema"
const findUp = require("find-up").sync
const Ajv = require("ajv")
const ajvSingleton = new Ajv()
const execSync = require("child_process").execSync

type AssertObjectOfTypeConfiguration = {
  filePath: string
  // https://github.com/vega/ts-json-schema-generator/issues/101#issuecomment-501321152
  // Each implementation has some shortcomings at the moment
  // "typescript-json-schema" -> can't generate circular types
  // "ts-json-schema-generator" -> more recent, but can't generated conditional or generic types
  implementation: "ts-json-schema-generator" | "typescript-json-schema"
}
/**
 * Takes any kind of object and verifies if it matches the target named
 * TypeScript type's jsonSchemaFromTSType.
 *
 * @param objectToValidate - the "untyped" object.
 * @param typeName - the target TypeScript type it supposedly is.
 * @param object: AssertObjectOfTypeConfiguration - extra configuration.
 * @return {boolean} [TODO:description]
 */
export function assertObjectOfType(
  objectToValidate: object,
  typeName: string,
  { filePath, implementation }: AssertObjectOfTypeConfiguration,
): boolean {
  if (!fs.existsSync) {
    throw new Error(`File ${filePath} does not exist`)
  }

  let jsonSchemaFromTSType: JSONSchema7 | undefined
  try {
    switch (implementation) {
      case "ts-json-schema-generator": {
        const generatorBinPath = path.resolve(
          __dirname,
          "./../node_modules/.bin/ts-json-schema-generator",
        )
        const result = execSync(
          `${generatorBinPath} --path '${filePath}.ts' --type '${typeName}' --expose all --no-top-ref`,
          { stdio: "pipe" },
        )
        jsonSchemaFromTSType = JSON.parse(result) as JSONSchema7
        break
      }
      case "typescript-json-schema": {
        const tsConfigPath =
          findUp(["tsconfig.json"], {
            cwd: path.dirname(filePath),
          }) || undefined
        const compilerOptions = ((tsConfigPath
          ? require(tsConfigPath)
          : {}) as unknown) as TJS.CompilerOptions

        const tsProgram = TJS.getProgramFromFiles([filePath])
        const maybeValidatedSchema = TJS.generateSchema(tsProgram, typeName)
        if (maybeValidatedSchema) {
          jsonSchemaFromTSType = maybeValidatedSchema as JSONSchema7
        }
        break
      }
    }
  } catch (err) {
    console.error(
      `It was not possible to generate the JSON schema for ${typeName} using '${implementation}'.`,
    )
    throw err
  }

  // validates that the untyped object is of the same type by comparing it's
  // structure to the jsonSchemaFromTSType at runtime
  const validate = ajvSingleton.compile(jsonSchemaFromTSType)
  const isValid = validate(objectToValidate)

  if (!isValid) {
    console.error(validate.errors)
    throw new Error(`Object supplied is not a valid ${typeName}`)
  }

  return true
}

export function isValidReduxModuleConfiguration(objectToValidate: object) {
  return assertObjectOfType(objectToValidate, "UserProvidedConfiguration", {
    filePath: path.resolve("./src/code_generator/redux/module.ts"),
    implementation: "typescript-json-schema",
  })
}

export function isValidTaggedComponentHierarchyConfiguration(
  objectToValidate: object,
) {
  return assertObjectOfType(objectToValidate, "Configuration", {
    filePath: path.resolve(
      "./src/code_generator/objects/tagged_component_hierarchy",
    ),
    implementation: "ts-json-schema-generator",
  })
}
