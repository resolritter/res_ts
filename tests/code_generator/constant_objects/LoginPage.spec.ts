import * as fs from "fs"
import { format } from "prettier"
import {
  generate,
  Configuration,
  Node,
} from "../../../src/code_generator/objects/tagged_component_hierarchy"
import { isValidTaggedComponentHierarchyConfiguration } from "../../../src/utils"
import { prettierConfigJson } from "../../utils"

const formFields: Node[] = [
  ["Label", []],
  ["Input", []],
]
const baseConfiguration: Configuration = {
  tree: [
    [
      "Login",
      [
        ["Email", formFields],
        ["Password", formFields],
      ],
    ],
  ],
}

describe("Test code generation", () => {
  it("should match the snapshot", () => {
    expect(
      format(generate(baseConfiguration), prettierConfigJson),
    ).toMatchSnapshot()
  })
})

describe("Test validation through JSON Schema", () => {
  it("should work for a schema configuration valid in TypeScript", () => {
    expect(
      isValidTaggedComponentHierarchyConfiguration(baseConfiguration),
    ).toBeTruthy()
  })

  it("should throw errors for a invalid schema", () => {
    // we don't want the errors to be print during test, so we disable it here and
    // restore it at the end
    const originalConsoleError = console.error
    console.error = jest.fn()

    expect(function () {
      isValidTaggedComponentHierarchyConfiguration({
        // this is incorrect because the tree is specified as an array
        tree: { name: "Root" },
      })
    }).toThrowErrorMatchingInlineSnapshot(
      `"Object supplied is not a valid Configuration"`,
    )

    expect(console.error).toHaveBeenCalled()
    console.error = originalConsoleError
  })
})
