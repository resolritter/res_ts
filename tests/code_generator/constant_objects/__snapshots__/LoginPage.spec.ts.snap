// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test code generation should match the snapshot 1`] = `
"module.exports = {
  Login: {
    id: \\"Login\\",
    Email: {
      id: \\"Login.Email\\",
      Label: {
        id: \\"Login.Email.Label\\"
      },
      Input: {
        id: \\"Login.Email.Input\\"
      }
    },
    Password: {
      id: \\"Login.Password\\",
      Label: {
        id: \\"Login.Password.Label\\"
      },
      Input: {
        id: \\"Login.Password.Input\\"
      }
    }
  }
};
"
`;
